### PR TITLE
feat: Copier Scaffolder Action

### PR DESCRIPTION
**Pull request by mspiegel31**
## Hey, I just made a Pull Request!
This is a scaffolder action to allow the creation of templates using the [copier](https://copier.readthedocs.io/en/stable/) template engine.  Copier is somewhat unique in the templating space, as it is version-aware.  This means that templates generated from a project are able to stay up-to-date with the template as they are modified over time.  Their update algorithm can be found [here](https://copier.readthedocs.io/en/stable/updating/#how-the-update-works)

### Screenshots
Screenshot 2023-10-04 at 12 03 20 PM

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by awanlin** [comment]
Hi @mspiegel31, just wanted to share this - https://github.com/backstage/backstage/issues/20266 - as it would be good to know the path forward before reviewing.

@freben can you provide some feedback, please?

**Comment by awanlin** [comment]
Also there already seems to be a module: https://github.com/telia-oss/scaffolder-backend-module-copier

**Comment by awanlin** [comment]
> @awanlin thanks for the heads-up! I think this would be great to have in backstage core, and this PR does add some extensions to `@backstage/backend-common`, so hopefully it's something that is worth looking at while #20266 is being worked through.

Let's wait to hear what @freben thinks 🤔

**Comment by mspiegel31** [comment]
> Also there already seems to be a module: https://github.com/telia-oss/scaffolder-backend-module-copier

that's true!  However, that module leverages the `fetchContents` function from `@backstage/plugin-scaffolder-backend`, which means that projects generated from templates won't have any version control information.  Since that's the "killer feature" of copier, it doesn't make much sense to use that particular action.

additionally, it does look there are just some copy-pasted pieces from the `cookiecutter` action in that codebase.  There is no such thing as a [copier.json](https://github.com/telia-oss/scaffolder-backend-module-copier/blob/main/src/actions/fetch/copier.ts#L46) file, for example; `copier` uses `yaml` for its configuration.

tl;dr I think this PR gives a more feature-complete solution, and does provide a way to get the functionality asked for in #14416 14416

**Comment by freben** [comment]
In the spirit of open source, it'd be really nice to contribute the improvements to the existing open source plugin, I feel. Maybe reach out to them?

**Comment by mspiegel31** [comment]
@freben I'll see what I can do.  This PR does contain some enhancements to `backend-common`; are contributions to that module also affected by #20266 or is it just net-new plugins?

**Comment by tudi2d** [comment]
Tja @tobias-estefors (main contributor of Telia's [scaffolder-backend-module-copier](https://github.com/telia-oss/scaffolder-backend-module-copier))! Would you be open for contribution on your scaffolder module? Would you be open to collaborate & review the changes? :)

**Comment by rdnkmrr** [comment]
@mspiegel31 Could you please guide/help me in creating the custom action for copier in backstage. I have explored the https://github.com/DiamondLightSource/scaffolder-backend-module-copier however that is not compatible with the latest backend system of backstage. We are trying to achieve the same in our organization as well.
